gogen/gogen_test.go

```diff
@@ -1843,6 +1843,156 @@ func (t *Container) ΛEnumTypeMap() map[string][]reflect.Type { return ΛEnumTyp
 func (*Container) ΛBelongingModule() string {
 	return "m1"
 }
+`,
+		},
+	}, {
+		name: "container with leaf setters",
+		inStructToMap: &ygen.ParsedDirectory{
+			Name: "Container",
+			Fields: map[string]*ygen.NodeDetails{
+				"leafStr": {
+					Name: "LeafStr",
+					YANGDetails: ygen.YANGNodeDetails{
+						Name:              "leafStr",
+						Defaults:          nil,
+						RootElementModule: "m1",
+						Path:              "/m1/foo/bar/leafStr",
+						LeafrefTargetPath: "",
+					},
+					Type: ygen.LeafNode,
+					LangType: &ygen.MappedType{
+						NativeType:        "string",
+						UnionTypes:        nil,
+						IsEnumeratedValue: false,
+						ZeroValue:         `""`,
+						DefaultValue:      nil,
+					},
+					MappedPaths:             [][]string{{"bar", "leafStr"}},
+					MappedPathModules:       [][]string{{"m1", "m1"}},
+					ShadowMappedPaths:       nil,
+					ShadowMappedPathModules: nil,
+				},
+				"leafUnion": {
+					Name: "LeafUnion",
+					YANGDetails: ygen.YANGNodeDetails{
+						Name:              "leafUnion",
+						Defaults:          nil,
+						RootElementModule: "m1",
+						Path:              "/m1/foo/bar/leafUnion",
+						LeafrefTargetPath: "",
+					},
+					Type: ygen.LeafNode,
+					LangType: &ygen.MappedType{
+						NativeType: "Container_U1_Union",
+						UnionTypes: map[string]ygen.MappedUnionSubtype{
+							"string": {
+								Index: 0,
+							},
+							"int8": {
+								Index: 1,
+							},
+						},
+						IsEnumeratedValue: false,
+						ZeroValue:         "nil",
+						DefaultValue:      nil,
+					},
+					MappedPaths:             [][]string{{"bar", "leafUnion"}},
+					MappedPathModules:       [][]string{{"m1", "m1"}},
+					ShadowMappedPaths:       nil,
+					ShadowMappedPathModules: nil,
+				},
+			},
+			Path:            "/m1/foo",
+			BelongingModule: "m1",
+		},
+		inGoOpts: GoOpts{
+			GenerateJSONSchema:  true,
+			GenerateLeafSetters: true,
+		},
+		want: wantGoStructOut{
+			structs: `
+// Container represents the /m1/foo YANG schema element.
+type Container struct {
+	LeafStr	*string	` + "`" + `path:"bar/leafStr" module:"m1/m1"` + "`" + `
+	LeafUnion	Container_U1_Union	` + "`" + `path:"bar/leafUnion" module:"m1/m1"` + "`" + `
+}
+
+// IsYANGGoStruct ensures that Container implements the yang.GoStruct
+// interface. This allows functions that need to handle this struct to
+// identify it as being generated by ygen.
+func (*Container) IsYANGGoStruct() {}
+`,
+			methods: `
+// SetLeafStr sets the value of the leaf LeafStr in the Container
+// struct.
+func (t *Container) SetLeafStr(v string) {
+	t.LeafStr = &v
+}
+
+// SetLeafUnion sets the value of the leaf LeafUnion in the Container
+// struct.
+func (t *Container) SetLeafUnion(v Container_U1_Union) {
```

Comment:
Not blocking in this PR, but I wonder whether in the future we should support `SetLeafUnion<Type>` which would also hide needing to create the wrapper type. This would be similar to the way that the proto library does `oneof` fields.

Reply:
Nice idea, unfortunately we can't easily do this right now because Go generics doesn't support methods whose parameters are generic because it complicates the interface system: https://go.dev/doc/faq#generic_methods